lib/loader/serialize/serial_segment.cpp

```diff
@@ -4,32 +4,194 @@ namespace WasmEdge {
+    return Unexpect(Res);
+  }
+  if (auto Res = serializeExpression(Seg.getExpr(), OutVec); unlikely(!Res)) {
+    spdlog::error(ErrInfo::InfoAST(ASTNodeAttr::Expression));
+    return Unexpect(Res);
+  }
+  return {};
 }
 
 // Serialize element segment. See "include/loader/serialize.h".
-void Serializer::serializeSegment(const AST::ElementSegment &,
-                                  std::vector<uint8_t> &) {
-  // Element segment: ...
-  // TODO
+Expect<void> Serializer::serializeSegment(const AST::ElementSegment &Seg,
+                                          std::vector<uint8_t> &OutVec) {
+  // Element segment: mode:u32 + tableidx:u32 + offset:expr + elemkind:reftype +
+  // vec(u32) + vec(expr)
+  std::vector<uint8_t> Result;
+  uint8_t Mode = 0x00;
+  switch (Seg.getMode()) {
+  case AST::ElementSegment::ElemMode::Declarative:
+    Mode |= 0x02;
+    [[fallthrough]];
+
+  case AST::ElementSegment::ElemMode::Passive:
+    Mode |= 0x01;
+    break;
+
+  default:
+    break;
+  }
+
+  // Serialize idx.
+  if (Seg.getIdx() != 0) {
+    serializeU32(Seg.getIdx(), Result);
+    if (Seg.getMode() == AST::ElementSegment::ElemMode::Active) {
+      Mode |= 0x02;
+    } else {
+      return logSerializeError(ErrCode::Value::Unreachable,
+                               ASTNodeAttr::Seg_Element);
+    }
+  }
+
+  // Serialize OffExpr.
+  if (Seg.getMode() == AST::ElementSegment::ElemMode::Active) {
+    if (auto Res = serializeExpression(Seg.getExpr(), Result); unlikely(!Res)) {
+      spdlog::error(ErrInfo::InfoAST(ASTNodeAttr::Expression));
+      return Unexpect(Res);
+    }
+  }
+
+  if (Seg.getRefType() == RefType::FuncRef) {
+    if (Mode != 0x00) {
+      // Serialize ElemKind.
+      Result.push_back(0x00);
+    }
+  } else {
+    // Serialize RefType.
+    if (auto Res = Conf.checkRefTypeProposals(Seg.getRefType(),
+                                              ASTNodeAttr::Seg_Element);
+        unlikely(!Res)) {
+      return Unexpect(Res);
+    }
+    Result.push_back(static_cast<uint8_t>(Seg.getRefType()));
+  }
+
+  // Distinguish between FuncIdx and Expr.
+  if (Seg.getInitExprs().size() != 0) {
+    auto IsExpr = false;
+    for (auto Expr : Seg.getInitExprs()) {
+      if (Expr.getInstrs().size() != 2 ||
+          Expr.getInstrs().at(0).getOpCode() != OpCode::Ref__func) {
+        IsExpr = true;
+        break;
+      }
+    }
+    if (IsExpr) {
+      Mode |= 0x04;
+    }
+  }
+
+  //  Mode > 0 cases are for BulkMemoryOperations or ReferenceTypes proposal.
+  if (Mode > 0 && !Conf.hasProposal(Proposal::BulkMemoryOperations) &&
+      !Conf.hasProposal(Proposal::ReferenceTypes)) {
+    return Conf.logNeedProposal(ErrCode::Value::ExpectedZeroByte,
+                                Proposal::BulkMemoryOperations,
+                                ASTNodeAttr::Seg_Element);
+  }
+
+  serializeU32(static_cast<uint32_t>(Seg.getInitExprs().size()), Result);
+  for (auto Expr : Seg.getInitExprs()) {
```

Comment:
Should we use `const auto &` instead for accessing the elements of each loop?
It looks like only reading the data from them.